src/System.Slices/src/System/SpanEnumerators.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System
+{
+    public partial struct Span<T>
+    {
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return new EnumeratorObject(this);
+        }
+
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+        {
+            return new EnumeratorObject(this);
+        }
+
+        /// <summary>
+        /// A struct-based enumerator, to make fast enumerations possible.
+        /// This isn't designed for direct use, instead see GetEnumerator.
+        /// </summary>
+        public struct Enumerator
+        {
```

Comment:
I have written some benchmarks to show why having two dedicated enumerators is better for performance. The code can be found here: https://github.com/adamsitnik/DemoCode/blob/master/Performance/Benchmarks/Collections/EnumeratorsBenchmarks.cs

The results (scroll down to summary): https://github.com/adamsitnik/DemoCode/blob/master/Performance/Benchmarks/results/EnumeratorsBenchmarks.txt

Summary:
1. Calling interface method on struct enumerators is very expensive. Looping with interface method calls compared to direct calls is x3 times slower on x86 and more than 10x slower on x64.
2.  On x86 looping with direct method calls to class enumerator is 15% faster than to direct with struct. But on the other hand on x64 the looping with direct calls to struct enumerator is more than 3x times faster than class.

That is why we have two dedicated enumerators here.